### PR TITLE
Task#2 set command

### PR DESCRIPTION
**Pull request by josephShokry**

### CODE REVIEW

File: source/models/index.py [hunk 1]
```diff
@@ -1,29 +1,33 @@
+from genericpath import isfile
 import os
 from outputs.exceptions import *
 
 class Index:
-    def __init__(self, table, index_value):
+    def __init__(self, table, index_name, index_value = None):
         self.table = table
```
Comment (mohamedabdelsalam) [inline source/models/index.py:6]: What is the difference between index name and index value ?

File: source/models/index.py [hunk 2]
```diff
@@ -1,29 +1,33 @@
...
     def get_path(self, index_value = None):
         if index_value == None:
-            return os.path.join(self.table.get_path(), os.path.join("indices" , self.index_value))
+            return os.path.join(self.table.get_path(), os.path.join("Indices" , self.index_name))
         else:
             return os.path.join(self.get_path(), self.index_value + ".txt")
 
-    def add_primary_key(self, primary_key, index_value):
-        old_primary_keys = set(self.get_primary_key(index_value))                
+    def add_primary_key(self, primary_key):
```
Comment (mohamedabdelsalam) [inline source/models/index.py]: It should take index value.

index is index name and metadata that belong to.

File: source/models/row.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+import os, json, uuid
+from outputs.exceptions import *
+from models.index import Index
+
+class Row:
```
Comment (mohamedabdelsalam) [inline source/models/row.py:5]: I reviewed this yesterday, why are u asking me to review this again ??

File: source/models/row.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            value = json.load(row_file)
+        return Row(table = table, value = value)
+
+    def delete(self):
+        self.__lock()
+        if not self.row_exists():
+            return
+        os.remove(self.get_path())
+        for index_name in self.table.table_metadata.index_keys:
+            index = Index(self.table, index_name = index_name, index_value = self.value[index_name])
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: Metadata has a list of index object. You shouldn’t create Index in anywhere else.
Index shouldn’t have the value. It take the value in method index.add_primary_key, index.remove_primary_key

File: source/models/index.py (hunk 1 above)
Comment (josephShokry) [reply source/models/index.py:6]: index name is like ```first name``` folder
but index value is like ```joseph.txt``` file that hold the PKs of students has joseph name

File: source/models/index.py (hunk 2 above)
Comment (josephShokry) [reply source/models/index.py]: it take it from the self.
as I create the object I give it the ```index value``` and store it at the self

File: source/models/row.py (hunk 3 above)
Comment (josephShokry) [reply source/models/row.py:5]: the problem was because the rebase i think it is done now

File: source/models/row.py (hunk 4 above)
Comment (josephShokry) [reply source/models/row.py]: ok now i get it, will work on it

File: source/models/database.py [hunk 5]
```diff
@@ -34,9 +36,8 @@ def __serialize_tables(self):
             table.serialize()
 
     def set(self, table_name, row):
-        # table = Table(table_name)
-        # table.set(row)
-        pass
+        table = Table(self, table_name = table_name)
```
Comment (mohamedabdelsalam) [inline source/models/database.py]: You already have a map of tables. You can get the table object from it. You don’t need to create a new one

File: source/models/index.py [hunk 6]
```diff
@@ -1,29 +1,33 @@
...
+        old_primary_keys = set(self.get_primary_key(self.index_value))                
         old_primary_keys.add(primary_key)
-        Index.__write_in_file(path = self.get_path(index_value), data = old_primary_keys)
-    
+        self.__write_in_file(data = old_primary_keys)
+
     def delete_primary_key(self, primary_key, index_value):
         old_primary_keys = self.get_primary_key(index_value)
-        old_primary_keys.remove(primary_key)
-        Index.__write_in_file(path = self.get_path(index_value), data = old_primary_keys)
```
Comment (mohamedabdelsalam) [inline source/models/index.py:26]: Same mistake you did before.
Don’t open same file in different method.
Create a method takes a list of primary_key

This method is doing the following 
- check if the list is empty, so delete the file
- Add overwrite the file with the list of primary key
- 
by other word use write_in_file method to do this and call it here

File: source/models/row.py (hunk 4 above)
Comment (josephShokry) [reply source/models/row.py]: done

File: source/models/table.py [hunk 7]
```diff
@@ -34,11 +37,32 @@ def get_path(self):
     def get_name(self):
         return self.__table_name
 
-    def set(self):
-        pass
+    def set(self, row):
```
Comment (mohamedabdelsalam) [inline source/models/table.py:40]: Move Validate in row constructor

File: source/models/table_metadata.py [hunk 8]
```diff
@@ -18,13 +18,18 @@ def get_path(self):
...
-        for index_value in self.index_keys: 
-            index = Index(table = self.table, index_value = index_value)
-            index.serialize()
\ No newline at end of file
+        os.makedirs(os.path.join(self.get_path(),"Indices"), exist_ok = True)
+        for index_name in self.index_keys: 
+            index = Index(table = self.table, index_name = index_name)
+            index.serialize()
+
+    def serialize_folders(self):
```
Comment (mohamedabdelsalam) [inline source/models/table_metadata.py]: You are creating folder. Serialize mean you convert an object in memory to text in file

File: source/models/index.py (hunk 1 above)
Comment (hussainmansour) [reply source/models/index.py:6]: same answer

File: source/models/table_metadata.py (hunk 8 above)
Comment (mohamedabdelsalam) [reply source/models/table_metadata.py]: Make it private

File: source/models/database.py (hunk 5 above)
Comment (josephShokry) [reply source/models/database.py]: ok done

File: source/models/table.py (hunk 7 above)
Comment (josephShokry) [reply source/models/table.py:40]: sorry my mistake, I did so but forget to remove it from here

File: source/models/index.py (hunk 6 above)
Comment (josephShokry) [reply source/models/index.py:26]: can you explain more

File: source/models/table.py [hunk 9]
```diff
@@ -38,12 +38,11 @@ def get_name(self):
         return self.__table_name
 
     def set(self, row):
-        self.__colomns_name_validate(row = row)
         row_obj = Row(table = self, value = row)
         if row_obj.row_exists() and self.table_metadata.enable_overwrite == "false":
             raise RowExists(message = "this row is already exists and can't overwrite")
         elif row_obj.row_exists():
-            old_row = Row(table = self).load_by_primary_key(table=self, primary_key = row_obj.primary_key)
+            old_row = Row(table = self).load_by_primary_key(table = self, primary_key = row_obj.primary_key)
```
Comment (mohamedabdelsalam) [inline source/models/table.py]: Load is a static method, why do u need to create an object to call it?
It should be Row.load

File: source/models/index.py (hunk 6 above)
Comment (mohamedabdelsalam) [reply source/models/index.py:26]: You should only write or delete file in the same method.
Rename write_in_file to update_in_file.

update_in_file should take an array of indices. 

If array is empty, delete the file
If not empty, overwrite the file by new values

File: source/models/row.py [hunk 10]
```diff
@@ -1,15 +1,19 @@
 import os, json, uuid
-import time
 from outputs.exceptions import *
 from models.index import Index
 
 class Row:
-    def __init__(self, table, value = None):
+    def __init__(self, table, value = {}):
         self.table = table
         self.value = value
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: Value should be private

File: source/models/row.py [hunk 11]
```diff
@@ -1,15 +1,19 @@
...
-        if self.table.table_metadata.primary_key not in value : 
-            value[self.table.table_metadata.primary_key] = str(uuid.uuid4().node)
-        self.primary_key = value[self.table.table_metadata.primary_key]
+        self.primary_key = self.__get_primary_key()
+
+    def __get_primary_key(self):
+        if self.table.table_metadata.primary_key not in self.value : 
+            self.value[self.table.table_metadata.primary_key] = str(uuid.uuid4().node)
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: You shouldn’t update value here, you should update this in constructor

File: source/models/row.py [hunk 12]
```diff
@@ -1,15 +1,19 @@
...
 
 class Row:
-    def __init__(self, table, value = None):
+    def __init__(self, table, value = {}):
         self.table = table
         self.value = value
-        if self.table.table_metadata.primary_key not in value : 
-            value[self.table.table_metadata.primary_key] = str(uuid.uuid4().node)
-        self.primary_key = value[self.table.table_metadata.primary_key]
+        self.primary_key = self.__get_primary_key()
```
Comment (mohamedabdelsalam) [inline source/models/row.py]: You don’t need primary_key instance variable. Get_primary_key() method is sufficient

File: source/models/table.py (hunk 9 above)
Comment (josephShokry) [reply source/models/table.py]: done

File: source/models/row.py (hunk 10 above)
Comment (josephShokry) [reply source/models/row.py]: why should it be private?
i think Hussin need this variable in the get

File: source/models/row.py (hunk 11 above)
Comment (mohamedabdelsalam) [reply source/models/row.py:13]: He should deal with row object. Value is hidden from him

File: source/models/row.py (hunk 10 above)
Comment (mohamedabdelsalam) [reply source/models/row.py]: He should deal with row object. Value is hidden from him

Comment (mohamedabdelsalam) [reply source/models/row.py]: I commented on Hussien PR that compare method should be in Row and rename It by has_attribute

Comment (mohamedabdelsalam) [reply source/models/row.py]: Read the advantage of encapsulating 
https://www.thoughtco.com/definition-of-encapsulation-958068#:~:text=Advantages%20of%20Encapsulation&text=Encapsulation%20protects%20an%20object%20from,the%20maintenance%20of%20the%20application

File: source/models/row.py (hunk 11 above)
Comment (josephShokry) [reply source/models/row.py]: i removed this function and made the check in the constructor.
done

File: source/models/row.py (hunk 12 above)
Comment (josephShokry) [reply source/models/row.py]: ok i will remove it

Comment (josephShokry) [reply source/models/row.py]: done

File: source/models/index.py (hunk 6 above)
Comment (josephShokry) [reply source/models/index.py:26]: ok i got it 
solved

File: source/models/index.py [hunk 13]
```diff
@@ -32,10 +32,13 @@ def get_primary_key(self, index_value):
...
+        self.add_primary_key(primary_key = primary_key, index_value = index_value)
 
-    def __write_in_file(self, data):
-        with open(self.get_index_value_path(self.index_value), mode ="w") as index_file:
-            index_file.write(",".join(data))
\ No newline at end of file
+    def __update_in_file(self, data, index_value):
+        if len(data) == 0 and os.path.isfile(self.get_path(index_value)) :
+            os.remove(self.get_path(index_value))
+        elif len(data)>0:
```
Comment (mohamedabdelsalam) [inline source/models/index.py]: Space before and after >

**Review by mohamedabdelsalam — APPROVED** [review]

**Review by hussainmansour — APPROVED** [review]

**Pull request merged**